collections/_posts/2022-11-14-DeformableConvolutionalNetworks.md

```diff
@@ -0,0 +1,130 @@
+However, there is no mechanism to help the model learn the geometric variations. The data augmentation is therefore limited by the user's knowledge. The proposed mechanism is end-to-end, adaptable to each architecture containing CNNs and allows CNNs to learn more complex deformations.
+
+&nbsp; 
+# Method
+
+## Basic convolution and limitation
+
+The following illustration is a simple convolution with a kernel size of $$ 3 \times 3$$ and a stride of $$1 \times 1$$. This calculation can be written as follows:
+
+$$ y(p_0) = \sum_{p_n \in R}w(p_n) \cdot x(p_0 + p_n) $$
+
+With $$p_n$$, The receptive field of the predefined convolution in the grid $$ R $$. In this specific case, the grid is equal to:
+
+$$ R = \{ (-1,-1), (-1, 0), (0, -1), ..., (0, 1), (1, 0), (1, 1) \} $$
+
+The grid $$ R $$ is static during training and inference. Thus the recepteive field of classical convolution is limited. To increase it, the kernel size, the pitch or the number of cascaded convolution can be increased.
+
+![](/collections/images/DeformableConvolutionalNetworks/standart_convolution.gif)
+
+## Deformable convolution
+
+In the proposed approach, the convolution operation can be rewritten as follows:
+
+$$ y(p_0) = \sum_{p_n \in R}w(p_n) \cdot x(p_0 + p_n + \Delta p_n) $$
+
+$$ \Delta p_n $$ correspond to the learnable grid. For $$ 3 \times 3$$ convolution, we have:
+
+$$ R = \{ (\Delta {x^1}, \Delta {y^1}), (\Delta {x^2}, \Delta {y^2})), ..., (\Delta {x^9}, \Delta {y^9}) \} $$
+
+In practice, additional layers are added to learn these components. For a $$ 3 \times 3 $$ convolution, 18 additional parameters are learned. The illustration below shows the mechanism:
+
+![](/collections/images/DeformableConvolutionalNetworks/deformable_convolution_illustration.jpg)
+
+An interesting point is that the sampling position is adaptable to the geometric transformation, especially in the sub-images below (c,d), where scaling and rotation have been applied.
+
+![](/collections/images/DeformableConvolutionalNetworks/sampling_location.jpg)
+
+In this scenario, the sample positions do not need to be integers. For fast calculation, the bilinear interpolation is implemented as follows:
+
+$$ x(p) = \sum_q max(0,1 - |p_x - q_x| ) \cdot max(0,1 - |p_y - q_y| )\cdot x(q)$$
+
+With $$ p = p_0 + p_n + \Delta p_n$$.
+&nbsp;
+
+## Deformable RoI Pooling (for Region Proposol Network)
+
+**RoI pooling**
+
+The same principle can be applied on pooling operation. Basic average RoI pooling is given by the following equation:
+
+$$ y(i,j) = \sum_{p \in bin(i, j)} \frac{ x(p_0 + p_n)}{n_{ij}} $$ 
+
+With $$bin(i,j)$$ the size of the grid surrounding the pixel at coordonate $$ (i, j) $$. Such as for deformable convolution, we add an extra learnable parameters:
```

Comment:
bin(i,j) is not the size of the grid (that would be n_{ij}), but rather the offsets for all coordinates in the bin.

Reply:
absolutely right!